plugins/async-node/core/src/index.test.ts

```diff
@@ -1,42 +1,120 @@
+import { AsyncNodePlugin, AsyncNodePluginPlugin } from './index';
 
 const basicFRFWithActions = {
-  id: "test-flow",
+  id: 'test-flow',
   views: [
     {
-      id: "my-view",
+      id: 'my-view',
       actions: [
         {
           asset: {
-            id: "action-0",
-            type: "action",
-            value: "{{foo.bar}}",
-          },
+            id: 'action-0',
+            type: 'action',
+            value: '{{foo.bar}}'
+          }
         },
         {
-          id: "uhh",
-          async: "true",
-        },
-      ],
-    },
+          id: 'uhh',
+          async: 'true'
+        }
+      ]
+    }
   ],
   navigation: {
-    BEGIN: "FLOW_1",
+    BEGIN: 'FLOW_1',
     FLOW_1: {
-      startState: "VIEW_1",
+      startState: 'VIEW_1',
       VIEW_1: {
-        state_type: "VIEW",
-        ref: "my-view",
-        transitions: {},
-      },
-    },
-  },
+        state_type: 'VIEW',
+        ref: 'my-view',
+        transitions: {}
+      }
+    }
+  }
 };
 
+const asyncNodeTest = async (resolvedValue: any, expectedActionType: string) => {
+  const plugin = new AsyncNodePlugin({
+    plugins: [new AsyncNodePluginPlugin()]
+  });
+
+  let deferredResolve: ((value: any) => void) | undefined;
+
+  plugin.hooks.onAsyncNode.tap('test', async (node: Node.Node) => {
+    return new Promise((resolve) => {
+      deferredResolve = resolve; // Promise would be resolved only once
+    });
+  });
+
+  let updateNumber = 0;
+
+  const player = new Player({ plugins: [plugin] });
+
+  let viewInstance : ViewInstance;
+  player.hooks.viewController.tap('async-node-test', (vc) => {
+    vc.hooks.view.tap('async-node-test', (view) => {
+      viewInstance = view;
+      view.hooks.onUpdate.tap('async-node-test', () => {
+        updateNumber++;
+      });
+    });
+  });
+
+  player.start(basicFRFWithActions as any);
+
+  let view = (player.getState() as InProgressState).controllers.view.currentView
+    ?.lastUpdate;
+
+  expect(view).toBeDefined();
+  expect(view?.actions[0].asset.type).toBe('action');
+  expect(view?.actions[1]).toBeUndefined();
+
+  await waitFor(() => {
+    expect(deferredResolve).toBeDefined();
+  });
+
+  // Consumer responds with null/undefined
+  if (deferredResolve) {
+    deferredResolve(resolvedValue);
+  }
+
+  await waitFor(() => {
+    expect(updateNumber).toBe(2);
+  });
+
+  view = (player.getState() as InProgressState).controllers.view.currentView
+    ?.lastUpdate;
+
+  expect(view?.actions[0].asset.type).toBe('action');
+  expect(view?.actions.length).toBe(1);
+  expect(updateNumber).toBe(2); // Replace with the actual expected number of updates
+
+  viewInstance.update();
+
+  await waitFor(() => {
+    expect(updateNumber).toBe(3);
+  });
+
+  view = (player.getState() as InProgressState).controllers.view.currentView
+    ?.lastUpdate;
+
+  expect(view?.actions[0].asset.type).toBe("action");
+  expect(view?.actions.length).toBe(1);
+  expect(updateNumber).toBe(3);
```

Comment:
you can remove this because it's implied [here](https://github.com/player-ui/player/pull/389/files#diff-fbc47a3e301bdf595a6d686ded538fe6767e534af8821c421f97dc56264d1332R99)

Reply:
done